README.md

```diff
@@ -487,6 +487,48 @@ static void BM_ManualTiming(benchmark::State& state) {
 BENCHMARK(BM_ManualTiming)->Range(1, 1<<17)->UseManualTime();
 ```
 
+There is also a second overload of `UseManualTime`, which allows the user to pass an arbitrary unit name, an optional scaling multiplier for those units, and a way to convert those units to a cost that will be used by Benchmark internally to determine maximum iterations/initial iteration cost and similar:
+
+```c++
+#include <cstddef>
+
+#if _WIN32
+#include <intrin.h>
+#else
+#include <x86intrin.h>
+#endif
+
+int64_t rdtsc() {
+  return static_cast<int64_t>(__rdtsc());
+}
+
+static double compute_cost(double measurement) {
+  // convert to some representative cost
+  return measurement;
+}
+
+static void BM_ManualTimingWithUnits(benchmark::State& state) {
+  int rotations = state.range(0);
+  std::unique_ptr<int32_t[]> data(new int32_t[rotations]]);
+
+  for (auto _ : state) {
+    auto start = rtdsc();
+
+    for (int i = 0; i < rotations; ++i) {
+      data[i] = _rotl(data[i], 16)
+    }
+
+    auto elapsed = rtdsc() - start;
+    state.SetIterationTime(static_cast<double>(elapsed));
+  }
+}
+BENCHMARK(BM_ManualTimingWithUnits)
+  ->Range(1, 1<<9)
+  ->UseManualTime("cycles", benchmark::kSeconds, compute_cost);
+```
+
+Internally, it treats the computed cost as "seconds", but this can be made to specify any arbitrary measurement unit for the benchmark. This will affect things like `MinTime`, so choose your limits carefully when specifying manual units that are not clearly convertible to seconds.
```

Comment:
Time is a well-defined unit of measure. It is not an arbitrary measurement unit for the benchmark. We shouldn't design an API around confusing the two.

Reply:
I can change the description to talk about it more appropriately...

Reply:
I don't think we should allow it to represent an arbitrary value to begin with.

If we do allow custom timers, I would like the timer to be provided to the library, and have the library invoke it using the same semantics it does with the other clocks. 

I would also like to move towards using strong types for times. e.g. `std::chrono::time_point` and `std::chrono::duration`. Then we don't need `BenchmarkTimeInfo`.

Reply:
I think that's fundamentally cross from what this PR is trying to do, which is allow custom measurements (e.g, with rdpmc and other instruction / cycle performance counters) to be printed out in benchmarking data.

While I generally agree with `std::chrono::time_point` and `std::chrono::duration` being the right choice, they're not possible to expose to the user because this is a C++03-interface library. But lifting the internals to do that is entirely possible.

Reply:
I don't really know what to do about this, really. If the PR isn't considered well-motivated enough (providing custom units to be used and reported by the benchmark) then I don't really have a way forward.

Reply:
The benchmark loop has been optimized within an inch of its life. It's as low-cost as possible; this is super important. Otherwise running the loop itself appears in or influences the timing data. This is BAD. It's bad bad not good. 

My main objection to this patch is that it makes you write benchmarks that make the main loop costly. This leads to bad benchmarks and bad benchmarking data.

I'm not saying your PR isn't well motivated: returning arbitrary data or measurements is a reasonable goal that we should support.

My objection is how it does this.

For example, the library automatically determines how many iterations to run the benchmark for by guessing at different iteration counts, and then seeing if the timing results meet a minimum threshold (Ex. if I run the benchmark 100 iterations, does it take 1 second).

In your manual clock example, do you want your "manual time" to be used to determine the number of iterations? Ie. If I say `--benchmark_min_time=2.0`, then benchmarks using your manual clock will run until your manual timer reports `2.0` units of w/e your manual clock represents?

At this point, I think I would be OK with an approach that allowed users to specify a manual timer that acts like the other timers. That is, it the timer should only only stopped/started when the rest of the timers are, and not every iteration.